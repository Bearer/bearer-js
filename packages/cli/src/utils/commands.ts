import BaseCommand from '../base-command'

export async function linkIntegration(this: BaseCommand, anIdentifier: string) {
  let identifier = anIdentifier
  let { integrationTitle } = this.bearerConfig as { integrationTitle: string }
  if (!identifier) {
    const { data } = await this.devPortalClient.request<{ integrations: Integration[] }>({ query: QUERY })
    if (!data.data) {
      throw 'Unable to fetch integration list'
    }

    const { integrations } = data.data!
    const { integration } = await this.inquirer.prompt<{ integration: Integration }>([
      {
        name: 'integration',
        type: 'list',
        choices: integrations.map(int => ({
          name: `${int.uuid} - ${int.name}`,
          value: int
        }))
      }
    ])
    integrationTitle = integration.name
    identifier = integration.uuid
  }

  const integrationRc = { integrationTitle, integrationId: identifier }
  this.bearerConfig.setIntegrationConfig(integrationRc)
  this.log('Integration successfully linked! ðŸŽ‰')
}

type Integration = {
  uuid: string
  name: string
  latestActivity?: {
    state: string
  }
}

const QUERY = `
query CLILinkIntegrationList {
  integrations(includeGloballyAvailable: false) {
    uuid: uuidv2
    name
    latestActivity {
      state
    }
  }
}
`
