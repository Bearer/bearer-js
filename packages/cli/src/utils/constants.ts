import { BearerEnv, BaseConfig } from '../types'

export const LOGIN_CLIENT_ID = process.env.BEARER_LOGIN_CLIENT_ID || 'Wgll39KqWnJWud473wq7hZhiXxeNjEU7'
export const BEARER_ENV = process.env.BEARER_ENV || 'production'
export const BEARER_AUTH_PORT = 45677
export const BEARER_LOGIN_PORT = 56789

export const CONFIGS: Record<BearerEnv, BaseConfig> = {
  dev: {
    IntegrationServiceHost: 'https://int.dev.bearer.sh/',
    IntegrationServiceUrl: 'https://int.dev.bearer.sh/api/v1/',
    DeveloperPortalAPIUrl: 'https://api.staging.bearer.sh/graphql',
    DeveloperPortalUrl: 'https://app.staging.bearer.sh/',
    CdnHost: 'https://static.dev.bearer.sh',
    BearerEnv: 'dev',
    LoginDomain: 'https://login.bearer.sh'
  },
  staging: {
    IntegrationServiceHost: 'https://int.staging.bearer.sh/',
    IntegrationServiceUrl: 'https://int.staging.bearer.sh/api/v1/',
    DeveloperPortalAPIUrl: 'https://api.staging.bearer.sh/graphql',
    DeveloperPortalUrl: 'https://app.staging.bearer.sh/',
    CdnHost: 'https://static.staging.bearer.sh',
    BearerEnv: 'staging',
    LoginDomain: 'https://login.bearer.sh'
  },
  production: {
    IntegrationServiceHost: 'https://int.bearer.sh/',
    IntegrationServiceUrl: 'https://int.bearer.sh/api/v1/',
    DeveloperPortalAPIUrl: 'https://api.bearer.sh/graphql',
    DeveloperPortalUrl: 'https://app.bearer.sh/',
    CdnHost: 'https://static.bearer.sh',
    BearerEnv: 'production',
    LoginDomain: 'https://login.bearer.sh'
  }
}

// tslint:disable max-line-length
export const SUCCESS_LOGIN_PAGE = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Authentication callback</title>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style>
      * {
        -webkit-font-smoothing: antialiased;
        -moz-osx-font-smoothing: grayscale;
      }
      html,
      body {
        background-color: #f5f7fb;
        font-family: 'Proxima Nova', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif,
          'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';
        text-align: center;
        font-size: 16px;
        line-height: 16px;
      }
      h1 {
        color: #00c682;
        font-size: 2rem;
        font-weight: 600;
        letter-spacing: 0.99px;
        line-height: 29px;
      }
      p {
        color: #343c5d;
        letter-spacing: 0.56px;
      }
      a {
        border-radius: 4px;
        display: inline-block;
        margin-top: 32px;
        padding: 12px 18px;
        background-color: #030d36;
        color: #ffffff;
        font-weight: 600;
        letter-spacing: 0.2px;
        text-decoration: none;
      }
      .outer {
        display: table;
        position: absolute;
        top: 0;
        left: 0;
        height: 100%;
        width: 100%;
      }
      .middle {
        display: table-cell;
        vertical-align: middle;
      }
      .inner {
        margin-left: auto;
        margin-right: auto;
        max-width: 700px;
      }
      .hint {
        font-size: 0.9rem;
        padding: 1.5rem;
        border: 1px solid #c2c9ea;
        border-radius: 4px;
        background-color: #ffffff;
        position: relative;
        top: 60px;
        margin-bottom: 60px;
      }
    </style>
  </head>
  <body>
    <div class="outer">
      <div class="middle">
        <div class="inner">
          <svg width="37" height="40" xmlns="http://www.w3.org/2000/svg">
            <path
              d="M.8 19.2h25.269l-6.635-6.634a.8.8 0 0 1 1.132-1.132l8 8a.8.8 0 0 1 0 1.132l-8 8a.8.8 0 0 1-1.132-1.132L26.07 20.8H.8a.8.8 0 0 1 0-1.6zm14.4 11.2a.8.8 0 0 1 .8.8v6.4a.8.8 0 0 0 .8.8h17.6a.8.8 0 0 0 .8-.8V2.4a.8.8 0 0 0-.8-.8H16.8a.8.8 0 0 0-.8.8v6.4a.8.8 0 0 1-1.6 0V2.4A2.4 2.4 0 0 1 16.8 0h17.6a2.4 2.4 0 0 1 2.4 2.4v35.2a2.4 2.4 0 0 1-2.4 2.4H16.8a2.4 2.4 0 0 1-2.4-2.4v-6.4a.8.8 0 0 1 .8-.8z"
              fill="#00C682"
              fill-rule="nonzero"
            />
          </svg>
          <h1>Successfully authenticated</h1>
          <p>You can close this window</p>
          <br />
        </div>
      </div>
    </div>
  </body>
</html>
`
