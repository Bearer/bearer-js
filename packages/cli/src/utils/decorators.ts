import Command from '../BaseCommand'

type Constructor<T> = new (...args: any[]) => T
type TCommand = InstanceType<Constructor<Command>>

export function RequireScenarioFolder() {
  return function(_target: any, _propertyKey: string | symbol, descriptor: PropertyDescriptor) {
    let originalMethod = descriptor.value
    descriptor.value = async function(this: TCommand) {
      if (this.bearerConfig.isScenarioLocation) {
        await originalMethod.apply(this, arguments)
      } else {
        this.error('This command must be run within a scenario folder.')
      }
    }
    return descriptor
  }
}

export function RequireLinkedScenario() {
  return function(_target: any, _propertyKey: string | symbol, descriptor: PropertyDescriptor) {
    let originalMethod = descriptor.value
    descriptor.value = async function(this: TCommand) {
      if (this.bearerConfig.hasScenarioLinked) {
        await originalMethod.apply(this, arguments)
      } else {
        const error =
          this.colors.bold('You scenario must be linked before running this command\n') +
          this.colors.yellow(this.colors.italic('Please run: bearer link'))
        this.error(error)
      }
    }
    return descriptor
  }
}

export function ensureFreshToken() {
  return function(_target: any, _propertyKey: string | symbol, descriptor: PropertyDescriptor) {
    let originalMethod = descriptor.value
    descriptor.value = async function(this: TCommand) {
      const { authorization, ExpiresAt } = this.bearerConfig.bearerConfig

      if (authorization && authorization.AuthenticationResult) {
        try {
          if (ExpiresAt < Date.now()) {
            this.ux.action.start('Refreshing token')
            await refreshMyToken(this)
            this.ux.action.stop()
          }
        } catch (error) {
          this.ux.action.stop(`Failed`)
          this.error(error.message)
        }
        // TS is complaining with TS2445, commenting out this for now
        // this.debug('Running original method')
        await originalMethod.apply(this, arguments)
        return descriptor
      } else {
        const error =
          this.colors.bold('⚠️ It looks like you are not logged in\n') +
          this.colors.yellow(this.colors.italic('Please run: bearer login'))
        this.error(error)
        return descriptor
      }
    }
    return descriptor
  }
}

async function refreshMyToken(command: TCommand): Promise<boolean | Error> {
  const { RefreshToken } = command.bearerConfig.bearerConfig.authorization.AuthenticationResult!

  // try {
  const { statusCode, body } = await command.serviceClient.refresh({ RefreshToken })

  switch (statusCode) {
    case 200: {
      const {
        authorization: { AuthenticationResult }
      } = body

      await command.bearerConfig.storeBearerConfig({
        ExpiresAt: Date.now() + AuthenticationResult.ExpiresIn * 1000,
        authorization: {
          AuthenticationResult: {
            ...AuthenticationResult,
            RefreshToken
          }
        }
      })
      return true
    }
    case 401: {
      throw new UnauthorizedRefreshTokenError()
    }
    default: {
      throw new UnexpectedRefreshTokenError(JSON.stringify(body, undefined, 2))
    }
  }
}

class UnauthorizedRefreshTokenError extends Error {
  constructor() {
    super('Something went wrong, please run bearer login and try again')
  }
}

class UnexpectedRefreshTokenError extends Error {
  constructor(body: any) {
    super(`body : ${body}`)
  }
}
