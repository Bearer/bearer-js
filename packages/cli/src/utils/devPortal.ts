import axios from 'axios'
import BaseCommand from '../base-command'

export const devPortalClient = (command: BaseCommand) => {
  const instance = axios.create({
    baseURL: command.constants.DeveloperPortalAPIUrl
  })
  instance.interceptors.response.use(
    r => r,
    error => {
      if (error.response && error.response.status) {
        command.error('Unauthorized action, please run bearer login first')
      }
      return Promise.reject(error)
    }
  )

  return {
    request: async <DataReturned>(data: { query: string; variables?: any }) => {
      const token = await command.bearerConfig.getToken()
      if (token) {
        return instance.post<{ data?: DataReturned }>('', data, {
          headers: {
            Authorization: `Bearer ${token.access_token}`
          }
        })
      }
      throw new UnauthorizedError()
    }
  }
}

export class UnauthorizedError extends Error {
  constructor() {
    super('Unauthorized request')
  }
}
