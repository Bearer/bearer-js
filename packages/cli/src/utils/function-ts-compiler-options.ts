import * as ts from 'typescript'
const options: ts.CompilerOptions = {
  allowUnreachableCode: false,
  declaration: false,
  noUnusedLocals: false,
  noUnusedParameters: false,
  allowSyntheticDefaultImports: true,
  experimentalDecorators: true,
  moduleResolution: ts.ModuleResolutionKind.NodeJs,
  module: ts.ModuleKind.CommonJS,
  target: ts.ScriptTarget.ES2017
}
export default options
