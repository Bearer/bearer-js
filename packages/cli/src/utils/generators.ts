import * as intents from '@bearer/intents'
import * as fs from 'fs-extra'
import * as globby from 'globby'
import * as path from 'path'
import * as ts from 'typescript'
import * as TJS from 'typescript-json-schema'

const config = ts.readConfigFile(path.join(__dirname, '../../templates/start', 'tsconfig.json'), ts.sys.readFile)

const NON_INTENT_NAMES = ['DBClient']
const INTENT_NAMES = Object.keys(intents).filter(intentName => !NON_INTENT_NAMES.includes(intentName))
const INTENT_TYPE_IDENTIFIER = 'intentType'
// const FUNCTION_NAME_IDENTIFIER = 'action'

const intentEntries: IIntentEntry[] = []

// TODO: CORE-197
// function getDefinition(tsType: ts.TypeNode, generator: TJS.JsonSchemaGenerator): TJS.Definition {
//   switch (tsType.kind) {
//     case ts.SyntaxKind.TypeReference:
//       try {
//         const name = ((tsType as ts.TypeReferenceNode).typeName as ts.Identifier).escapedText.toString()
//         return generator.getSchemaForSymbol(name) as any
//       } catch (e) {
//         // TODO: re-use type reference
//         console.debug(e)
//         return { type: 'any' }
//       }
//     case ts.SyntaxKind.TypeLiteral: {
//       const typeNode = tsType as ts.TypeLiteralNode
//       return {
//         type: 'object',
//         properties: {
//           ...typeNode.members.reduce((acc, m) => {
//             const member = m as ts.PropertySignature
//             const name = (member.name as ts.Identifier).escapedText.toString()
//             return {
//               ...acc,
//               [name]: {
//                 description: name,
//                 name,
//                 required: !member.questionToken,
//                 schema: getDefinition(member.type!, generator)
//               }
//             }
//           }, {})
//         }
//       }
//     }
//     case ts.SyntaxKind.NumberKeyword:
//       return { type: 'number' }
//     case ts.SyntaxKind.BooleanKeyword:
//       return { type: 'boolean' }
//     case ts.SyntaxKind.StringKeyword:
//       return { type: 'string' }
//     default: {
//       return { type: 'object' } as any
//     }
//   }
// }

class IntentNodeAdapter implements IIntentEntry {
  constructor(
    readonly intentName: string,
    private readonly node: ts.ClassDeclaration,
    // @ts-ignore
    private readonly generator: TJS.JsonSchemaGenerator
  ) {}

  get intentClassName() {
    return getIdentifier(this.node).escapedText.toString()
  }

  get intentType() {
    return getPropertyValue(this.node, INTENT_TYPE_IDENTIFIER)
  }

  get paramsSchema() {
    const defaultParams = [...DEFAULT_PARAMS]
    // TODO: CORE-197, needs to be fixed to keep OpenApi spec generated correctly
    return defaultParams
    // const typeNode = getFunctionParameterType(this.node, FUNCTION_NAME_IDENTIFIER, 'params')
    // if (!typeNode) {
    //   return defaultParams
    // }
    // const paramsSchema = getDefinition(typeNode, this.generator)
    // return [...this.adaptParamsSchema(paramsSchema), ...defaultParams]
  }

  adaptParamsSchema({ properties = {} }: { properties?: { [key: string]: any } }): ISchemaParam[] {
    if (properties) {
      return Object.keys(properties).map(propName => {
        const { schema, required, name, description } = properties[propName]
        return {
          schema: schema || { type: 'string' },
          in: 'query',
          description: description || propName,
          required: required !== undefined ? required : true,
          name: name || propName
        } as ISchemaParam
      })
    }
    return []
  }

  get bodySchema() {
    // TODO: CORE-197, needs to be fixed to keep OpenApi spec generated correctly
    return {}
    // const typeNode = getFunctionParameterType(this.node, FUNCTION_NAME_IDENTIFIER, 'body')
    // if (!typeNode) {
    //   return {}
    // }
    // return getDefinition(typeNode, this.generator)
  }

  get outputSchema() {
    return {}
  }

  get adapt(): IIntentEntry {
    return {
      intentClassName: this.intentClassName,
      intentName: this.intentName,
      intentType: this.intentType,
      paramsSchema: this.paramsSchema,
      bodySchema: this.bodySchema,
      outputSchema: this.outputSchema
    }
  }
}

export function isIntentClass(tsNode: ts.Node): boolean {
  const isClass = ts.isClassDeclaration(tsNode) && tsNode.name
  if (!isClass) {
    return false
  }
  const intentTypeValue = getPropertyValue(tsNode as ts.ClassDeclaration, INTENT_TYPE_IDENTIFIER)
  return INTENT_NAMES.includes(intentTypeValue)
}

export function getIdentifier(tsNode: ts.ClassDeclaration | ts.PropertyDeclaration): ts.Identifier {
  return tsNode.name as ts.Identifier
}

export function getIntentName(tsSourceFile: ts.SourceFile): string {
  return path.basename(tsSourceFile.fileName).split('.')[0]
}

export function getPropertyValue(tsNode: ts.ClassDeclaration, propertyName: string): any {
  if (tsNode.members) {
    const declaration = tsNode.members.find(node => {
      return ts.isPropertyDeclaration(node) && (node.name as ts.Identifier).escapedText.toString() === propertyName
    }) as ts.PropertyDeclaration
    if (declaration && declaration.initializer) {
      return (declaration.initializer as ts.StringLiteral).text
    }
  }
}

// TODO: CORE-197
// function getFunctionParameterType(
//   tsNode: ts.ClassDeclaration,
//   functionNameIdentifier: string,
//   parameterIdentifier: string
// ): ts.TypeNode | undefined {
//   if (tsNode.members) {
//     const methodDeclaration = tsNode.members.find(node => {
//       return (
//         ts.isMethodDeclaration(node) && (node.name as ts.Identifier).escapedText.toString() ===
// functionNameIdentifier
//       )
//     }) as ts.MethodDeclaration

//     if (methodDeclaration.parameters) {
//       const parameter = methodDeclaration.parameters.find(node => {
//         if (!ts.isParameter(node)) {
//           return false
//         }
//         // @ts-
//         console.log('[BEARER]', 'node.name', node.getText())
//         return (node.name as ts.Identifier).escapedText.toString() === parameterIdentifier
//       })
//       if (parameter) {
//         return parameter.type
//       }
//     }
//   }
//   return
// }

export function transformer(generator: TJS.JsonSchemaGenerator): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext) => {
    return (tsSourceFile: ts.SourceFile) => {
      function visit(tsNode: ts.Node) {
        if (isIntentClass(tsNode)) {
          const adapter = new IntentNodeAdapter(getIntentName(tsSourceFile), tsNode as ts.ClassDeclaration, generator)
          intentEntries.push(adapter.adapt)
        }
        return tsNode
      }
      return ts.visitEachChild(tsSourceFile, visit, context)
    }
  }
}

export class IntentCodeProcessor {
  constructor(private readonly srcIntentsDir: string, private readonly transformer: any) {}

  async run() {
    const files = await globby(`${this.srcIntentsDir}/*.ts`)

    files.forEach(file => {
      const sourceFile = ts.createSourceFile(
        file,
        fs.readFileSync(file, 'utf8'),
        ts.ScriptTarget.Latest,
        false,
        ts.ScriptKind.TSX
      )
      ts.transform(sourceFile, [this.transformer])
    })
  }
}

export class OpenApiSpecGenerator {
  constructor(
    private readonly srcIntentsDir: string,
    private readonly bearerConfig: { scenarioTitle: string | undefined; scenarioUuid: string }
  ) {}

  async build() {
    const files = await globby(`${this.srcIntentsDir}/*.ts`)
    const programGenerator = TJS.getProgramFromFiles(
      files,
      {
        ...config.config.compilerOptions,
        allowUnusedLabels: true,
        // be indulgent
        noUnusedParameters: false,
        noUnusedLocals: false
      },
      './ok'
    )

    const generator = TJS.buildGenerator(programGenerator, {
      required: true
    })

    if (!generator) {
      throw new Error('Please fix above issues before')
    }

    files.forEach(file => {
      const sourceFile = ts.createSourceFile(
        file,
        fs.readFileSync(file, 'utf8'),
        ts.ScriptTarget.Latest,
        false,
        ts.ScriptKind.TSX
      )
      ts.transform(sourceFile, [transformer(generator)])
    })
    return this.generate(intentEntries, this.bearerConfig)
  }

  generate(
    entries: IIntentEntry[],
    { scenarioTitle, scenarioUuid }: { scenarioTitle?: string; scenarioUuid: string }
  ): IOpenApiSpec {
    return {
      openapi: '3.0.0',
      info: {
        description: `openapi definition file for ${scenarioTitle}`,
        version: '0.0.1',
        title: scenarioTitle,
        contact: {
          email: 'bearer@bearer.sh'
        },
        license: {
          name: 'MIT'
        }
      },
      servers: [
        {
          url: 'https://int.bearer.sh/backend/api/v1'
        }
      ],
      tags: [
        {
          name: 'integration',
          description: `List of endpoints providing backend to backend integration with ${scenarioTitle}`,
          externalDocs: {
            description: 'Find out more',
            url: 'https://www.bearer.sh'
          }
        }
      ],
      paths: entries.reduce((acc, entry) => {
        return {
          ...acc,
          [`/${scenarioUuid}/${entry.intentName}`]: {
            post: {
              parameters: entry.paramsSchema,
              summary: entry.intentName,
              requestBody: {
                content: {
                  'application/json': {
                    schema: entry.bodySchema
                  }
                }
              },
              responses: {
                '401': {
                  description: 'Access forbidden',
                  content: {
                    'application/json': {
                      schema: {
                        type: 'object',
                        properties: {
                          error: {
                            type: 'string'
                          }
                        }
                      }
                    }
                  }
                },
                '403': {
                  description: 'Unauthorized',
                  content: {
                    'application/json': {
                      schema: {
                        type: 'object',
                        properties: {
                          error: {
                            type: 'string'
                          }
                        }
                      }
                    }
                  }
                },
                '200': {
                  description: 'Share',
                  content: {
                    'application/json': {
                      schema: entry.outputSchema
                    }
                  }
                }
              }
            }
          }
        }
      }, {})
    }
  }
}

const DEFAULT_PARAMS: ISchemaParam[] = [
  {
    in: 'header',
    name: 'authorization',
    schema: {
      type: 'string'
    },
    description: 'API Key',
    required: true
  },
  {
    name: 'authId',
    schema: {
      type: 'string',
      format: 'uuid'
    },
    in: 'query',
    description: 'User Identifier',
    required: true
  }
]

export interface ISchemaParam {
  in: 'query' | 'header'
  schema: {
    type: 'string'
    format?: 'uuid'
  }
  description: string
  required: boolean
  name: string
}

export interface IOpenApiSpec {
  openapi: string
  info: any
  servers: any[]
  tags: any[]
  paths: {
    [key: string]: {
      post: {
        parameters: any[]
        summary: any[]
        requestBody: any[]
        responses: any[]
      }
    }
  }
}

interface IIntentEntry {
  intentClassName: string
  intentType: string
  intentName: string
  paramsSchema: ISchemaParam[]
  bodySchema: any
  outputSchema: any
}
