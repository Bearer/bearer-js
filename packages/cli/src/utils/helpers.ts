import * as copy from 'copy-template-dir'
import * as path from 'path'

import BaseCommand from '../BaseCommand'

export function copyFiles(
  command: BaseCommand,
  sourceDirectory: string,
  outPutDirectory: string,
  vars: any
): Promise<boolean | string> {
  const base = process.cwd()
  function printFiles(files: Array<string>) {
    files.forEach(file => {
      command.log(command.colors.gray(`    create: `) + command.colors.white(file.replace(base + '/', '')))
    })
  }

  return new Promise((resolve, reject) => {
    copy(path.join(__dirname, '..', '..', 'templates', sourceDirectory), outPutDirectory, vars, (err, createdFiles) => {
      if (err) {
        reject(err)
      } else {
        printFiles(createdFiles)
        resolve(true)
      }
    })
  })
}
