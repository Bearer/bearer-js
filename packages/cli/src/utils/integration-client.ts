import axios, { AxiosInstance } from 'axios'

import BaseCommand from '../base-command'

export class IntegrationClient {
  private client: AxiosInstance

  constructor(baseURL: string, authorization?: string, version?: string) {
    const headers = {
      Authorization: authorization,
      ['BEARER-CLI-VERSION']: version
    }
    this.client = axios.create({
      headers,
      baseURL
    })
  }

  async getIntegrationArchiveUploadUrl(orgId: string, integrationId: string): Promise<string> {
    try {
      const { data } = await this.client.post('integration-urls', { orgId, integrationId })
      return data.url
    } catch (e) {
      throw e
    }
  }
}

export default (command: BaseCommand): IntegrationClient => {
  return new IntegrationClient(
    command.bearerConfig.DeploymentUrl,
    command.bearerConfig.bearerConfig.authorization.AuthenticationResult!.IdToken,
    command.config.version
  )
}
