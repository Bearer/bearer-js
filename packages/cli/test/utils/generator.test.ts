import * as path from 'path'

import { OpenApiSpecGenerator } from '../../src/utils/generators'

describe('generators', () => {
  it('exports Spec Generator', () => {
    expect(OpenApiSpecGenerator).toBeTruthy()
  })
  // TODO: CORE-197

  describe('build', () => {
    it('generates openapi documentation', async () => {
      const generator = new OpenApiSpecGenerator(path.join(__dirname, '__fixtures__/generators'), {
        integrationTitle: 'test',
        integrationUuid: '123-test'
      })
      const build = await generator.build()
      expect(build).toMatchSnapshot()
    })
  })
})
