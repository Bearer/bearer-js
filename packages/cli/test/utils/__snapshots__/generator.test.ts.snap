// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generators build generates openapi documentation 1`] = `
"{
  \\"openapi\\": \\"3.0.0\\",
  \\"info\\": {
    \\"description\\": \\"openapi definition file for test\\",
    \\"version\\": \\"0.0.1\\",
    \\"title\\": \\"test\\",
    \\"contact\\": {
      \\"email\\": \\"bearer@bearer.sh\\"
    },
    \\"license\\": {
      \\"name\\": \\"MIT\\"
    }
  },
  \\"servers\\": [
    {
      \\"url\\": \\"https://int.bearer.sh/api/v2/functions/backend/\\"
    }
  ],
  \\"tags\\": [
    {
      \\"name\\": \\"integration\\",
      \\"description\\": \\"List of endpoints providing backend to backend integration with test\\",
      \\"externalDocs\\": {
        \\"description\\": \\"Find out more\\",
        \\"url\\": \\"https://www.bearer.sh\\"
      }
    }
  ],
  \\"paths\\": {
    \\"/123-test/object-literal-type\\": {
      \\"post\\": {
        \\"parameters\\": [
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"authorization\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"description\\": \\"API Key\\",
            \\"required\\": true
          },
          {
            \\"name\\": \\"authId\\",
            \\"schema\\": {
              \\"type\\": \\"string\\",
              \\"format\\": \\"uuid\\"
            },
            \\"in\\": \\"query\\",
            \\"description\\": \\"User Identifier\\",
            \\"required\\": true
          }
        ],
        \\"summary\\": \\"object-literal-type\\",
        \\"requestBody\\": {
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"params\\": {
                    \\"type\\": \\"object\\",
                    \\"properties\\": {
                      \\"inlineParam\\": {
                        \\"type\\": \\"string\\"
                      },
                      \\"stringEnum\\": {},
                      \\"inlineNumber\\": {
                        \\"type\\": \\"number\\"
                      },
                      \\"nestedObject\\": {
                        \\"type\\": \\"object\\",
                        \\"properties\\": {
                          \\"name\\": {
                            \\"type\\": \\"string\\"
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        },
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"object-literal-type\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"data\\": {
                      \\"type\\": \\"object\\",
                      \\"properties\\": {
                        \\"expectedData\\": {
                          \\"type\\": \\"array\\",
                          \\"items\\": {
                            \\"type\\": \\"string\\"
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          },
          \\"401\\": {
            \\"description\\": \\"Access forbidden\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"error\\": {
                      \\"type\\": \\"string\\"
                    }
                  }
                }
              }
            }
          },
          \\"403\\": {
            \\"description\\": \\"Unauthorized\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"error\\": {
                      \\"type\\": \\"string\\"
                    }
                  }
                }
              }
            }
          }
        }
      }
    },
    \\"/123-test/type-alias\\": {
      \\"post\\": {
        \\"parameters\\": [
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"authorization\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"description\\": \\"API Key\\",
            \\"required\\": true
          },
          {
            \\"name\\": \\"authId\\",
            \\"schema\\": {
              \\"type\\": \\"string\\",
              \\"format\\": \\"uuid\\"
            },
            \\"in\\": \\"query\\",
            \\"description\\": \\"User Identifier\\",
            \\"required\\": true
          }
        ],
        \\"summary\\": \\"type-alias\\",
        \\"requestBody\\": {
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"aliasParam\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"stringEnum\\": {},
                  \\"inlineNumber\\": {
                    \\"type\\": \\"number\\"
                  },
                  \\"nestedObject\\": {
                    \\"type\\": \\"object\\",
                    \\"properties\\": {
                      \\"name\\": {
                        \\"type\\": \\"string\\"
                      }
                    }
                  }
                }
              }
            }
          }
        },
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"type-alias\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {}
                }
              }
            }
          },
          \\"401\\": {
            \\"description\\": \\"Access forbidden\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"error\\": {
                      \\"type\\": \\"string\\"
                    }
                  }
                }
              }
            }
          },
          \\"403\\": {
            \\"description\\": \\"Unauthorized\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"error\\": {
                      \\"type\\": \\"string\\"
                    }
                  }
                }
              }
            }
          }
        }
      }
    },
    \\"/123-test/undefined-params-and-return\\": {
      \\"post\\": {
        \\"parameters\\": [
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"authorization\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"description\\": \\"API Key\\",
            \\"required\\": true
          }
        ],
        \\"summary\\": \\"undefined-params-and-return\\",
        \\"requestBody\\": {
          \\"content\\": {
            \\"application/json\\": {}
          }
        },
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"undefined-params-and-return\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"data\\": {}
                  }
                }
              }
            }
          },
          \\"401\\": {
            \\"description\\": \\"Access forbidden\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"error\\": {
                      \\"type\\": \\"string\\"
                    }
                  }
                }
              }
            }
          },
          \\"403\\": {
            \\"description\\": \\"Unauthorized\\",
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"error\\": {
                      \\"type\\": \\"string\\"
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}"
`;
